File: backend/mr-db-api/src/main/java/io/featurehub/db/api/ApplicationApi.java
package io.featurehub.db.api;

import io.featurehub.mr.model.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
import java.util.Set;
import java.util.UUID;

public interface ApplicationApi {

  @Nullable
  ApplicationSummary getApplicationSummary(@NotNull UUID appId);

  class DuplicateApplicationException extends Exception {}
  class DuplicateFeatureException extends Exception {}

  @Nullable
  Application createApplication(@NotNull UUID portfolioId, @NotNull CreateApplication application, @NotNull Person current) throws DuplicateApplicationException;

  @NotNull
  List<Application> findApplications(@NotNull UUID portfolioId, @Nullable String filter, SortOrder order,
                                     @NotNull Opts opts, @NotNull Person current, boolean loadAll);

  boolean deleteApplication(@NotNull UUID portfolioId, UUID applicationId);

  @Nullable Application getApplication(@NotNull UUID appId, @NotNull Opts opts);

  @Nullable Application updateApplication(@NotNull UUID appId, @NotNull Application application, @NotNull Opts opts) throws DuplicateApplicationException, OptimisticLockingException;

  @Nullable List<Feature> createApplicationFeature(@NotNull UUID appId, Feature feature, Person person, @NotNull Opts opts) throws DuplicateFeatureException;
  @Nullable List<Feature> updateApplicationFeature(@NotNull UUID appId, String key, Feature feature, @NotNull Opts opts) throws DuplicateFeatureException, OptimisticLockingException;
  @Nullable List<Feature> getApplicationFeatures(@NotNull UUID appId, @NotNull Opts opts);
  @Nullable List<Feature> deleteApplicationFeature(@NotNull UUID appId, String key);
  @Nullable Feature getApplicationFeatureByKey(@NotNull UUID appId, @NotNull String key, @NotNull Opts opts);

  @NotNull Set<UUID> findFeatureEditors(@NotNull UUID appId);
  boolean personIsFeatureEditor(@NotNull UUID appId, @NotNull UUID personId);

  @NotNull ApplicationPermissions findApplicationPermissions(@NotNull UUID appId, @NotNull UUID personId);

  /**
   * Those who can create features
   * @param appId
   * @return
   */
  @NotNull Set<UUID> findFeatureCreators(@NotNull UUID appId);
  boolean personIsFeatureCreator(@NotNull UUID appId, @NotNull UUID personId);

  @NotNull Set<UUID> findFeatureReaders(@NotNull UUID appId);
  boolean personIsFeatureReader(UUID appId, UUID personId);
}

File: backend/mr-db-api/src/main/kotlin/io/featurehub/db/api/ApplicationApi.kt
package io.featurehub.db.api

import io.featurehub.mr.model.*
import java.util.*

interface ApplicationApi {
  fun getApplicationSummary(appId: UUID): ApplicationSummary?
  class DuplicateApplicationException : Exception()
  class DuplicateFeatureException : Exception()

  @Throws(DuplicateApplicationException::class)
  fun createApplication(portfolioId: UUID, application: CreateApplication, current: Person): Application?
  fun findApplications(
    portfolioId: UUID, filter: String?, order: SortOrder?,
    opts: Opts, current: Person, loadAll: Boolean
  ): List<Application>

  fun deleteApplication(portfolioId: UUID, applicationId: UUID): Boolean
  fun getApplication(appId: UUID, opts: Opts): Application?

  @Throws(DuplicateApplicationException::class, OptimisticLockingException::class)
  fun updateApplication(appId: UUID, application: Application, opts: Opts): Application?
  fun updateApplicationOnPortfolio(portfolioId: UUID, application: Application, opts: Opts): Application?

  @Throws(DuplicateFeatureException::class)
  fun createApplicationFeature(appId: UUID, createFeature: CreateFeature, person: Person, opts: Opts): List<Feature>

  @Throws(DuplicateFeatureException::class, OptimisticLockingException::class)
  fun updateApplicationFeature(appId: UUID, key: String, feature: Feature, opts: Opts): List<Feature>?

  @Throws(DuplicateFeatureException::class, OptimisticLockingException::class)
  fun updateApplicationFeature(appId: UUID, feature: Feature, opts: Opts): List<Feature>?

    fun getApplicationFeatures(appId: UUID, opts: Opts): List<Feature>
  fun deleteApplicationFeature(appId: UUID, key: String): List<Feature>?
  fun getApplicationFeatureByKey(appId: UUID, key: String, opts: Opts): Feature?
  fun findFeatureEditors(appId: UUID): Set<UUID>
  fun personIsFeatureEditor(appId: UUID, personId: UUID): Boolean
  fun findApplicationPermissions(appId: UUID, personId: UUID): ApplicationPermissions

  /**
   * Those who can create features
   * @param appId
   * @return
   */
  fun findFeatureCreators(appId: UUID): Set<UUID>
  fun personIsFeatureCreator(appId: UUID, personId: UUID): Boolean
  fun findFeatureReaders(appId: UUID): Set<UUID>
  fun personIsFeatureReader(appId: UUID, personId: UUID): Boolean
}
